### PR TITLE
[pytorch] | [training] | [build] | [test] | [benchmark] | [sagemaker]Adding sm-model-parallel pytorch to 2.0.1

### PR DESCRIPTION
**Pull request by access2rohit**
*GitHub Issue #, if available:*

**Note**: 
- If merging this PR should also close the associated Issue, please also add that Issue # to the Linked Issues section on the right. 

- All PR's are checked weekly for staleness. This PR will be closed if not updated in 30 days.

### Description
SM Model Parallel v2

### Tests run

**NOTE: By default, docker builds are disabled. In order to build your container, please update dlc_developer_config.toml and specify the framework to build in "build_frameworks"**
- [ ] I have run builds/tests on commit <INSERT COMMIT ID> for my changes.

**NOTE: If you are creating a PR for a new framework version, please ensure success of the standard, rc, and efa sagemaker remote tests by updating the dlc_developer_config.toml file:**

- [ ] Revision A: `sagemaker_remote_tests = "standard"`
- [ ] Revision B: `sagemaker_remote_tests = "rc"`
- [ ] Revision C: `sagemaker_remote_tests = "efa"`

**Additionally, please run the sagemaker local tests in at least one revision:**
- [ ] `sagemaker_local_tests = true`

### Formatting
- [ ] I have run `black -l 100` on my code (formatting tool: https://black.readthedocs.io/en/stable/getting_started.html)

## PR Checklist
- [x] I've prepended PR tag with frameworks/job this applies to : [mxnet, tensorflow, pytorch] | [ei/neuron/graviton] | [build] | [test] | [benchmark] | [ec2, ecs, eks, sagemaker]
- [ ] If the PR changes affects SM test, I've modified dlc_developer_config.toml in my PR branch by setting sagemaker_tests = true and efa_tests = true
- [ ] If this PR changes existing code, the change fully backward compatible with pre-existing code. (Non backward-compatible changes need special approval.)
- [ ] (If applicable) I've documented below the DLC image/dockerfile this relates to
- [ ] (If applicable) I've documented below the tests I've run on the DLC image
- [ ] (If applicable) I've reviewed the licenses of updated and new binaries and their dependencies to make sure all licenses are on the Apache Software Foundation Third Party License Policy Category A or Category B license list.  See [https://www.apache.org/legal/resolved.html](https://www.apache.org/legal/resolved.html).
- [ ] (If applicable) I've scanned the updated and new binaries to make sure they do not have vulnerabilities associated with them.

## Pytest Marker Checklist
- [ ] (If applicable) I have added the marker `@pytest.mark.model("<model-type>")` to the new tests which I have added, to specify the Deep Learning model that is used in the test (use `"N/A"` if the test doesn't use a model)
- [ ] (If applicable) I have added the marker `@pytest.mark.integration("<feature-being-tested>")` to the new tests which I have added, to specify the feature that will be tested
- [ ] (If applicable) I have added the marker `@pytest.mark.multinode(<integer-num-nodes>)` to the new tests which I have added, to specify the number of nodes used on a multi-node test
- [ ] (If applicable) I have added the marker `@pytest.mark.processor(<"cpu"/"gpu"/"eia"/"neuron">)` to the new tests which I have added, if a test is specifically applicable to only one processor type

#### EIA/NEURON/GRAVITON Testing Checklist
* When creating a PR:
- [ ] I've modified `dlc_developer_config.toml` in my PR branch by setting `ei_mode = true`, `neuron_mode = true` or `graviton_mode = true`

#### Benchmark Testing Checklist
* When creating a PR:
- [x] I've modified `dlc_developer_config.toml` in my PR branch by setting `benchmark_mode = true`

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license. I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: pytorch/training/docker/2.0/py3/cu118/Dockerfile.gpu [hunk 1]
```diff
@@ -429,6 +429,17 @@ RUN /opt/conda/bin/mamba install -y python=${PYTHON_VERSION} pytorch=${PYTORCH_V
...
+# Override only sm-pytorch
+RUN /opt/conda/bin/mamba install --force-reinstall -y \
+  pytorch=2.0.1=sm_py3.10_cuda11.8_cudnn8.9.5_nccl2.18.3_0_smp_2.0.0_pt_2.0.1 --override-channels \
+  -c https://sagemaker-distributed-model-parallel.s3.us-west-2.amazonaws.com/smp-2.0.0-pt-2.0.1/2023-12-11/smp-v2/ \
+  -c pytorch \
+  -c numba/label/dev \
+  -c pytorch-nightly \
+  -c nvidia \
+  -c conda-forge \
+  && /opt/conda/bin/mamba clean -afy
```
Comment (roywei) [inline pytorch/training/docker/2.0/py3/cu118/Dockerfile.gpu:441]: Does the ofi package comes with this installation? if not, you can manually install from our conda channel. `mamba install aws-ofi-nccl --strict-channel-priority --override-channels -c https://aws-ml-conda.s3.us-west-2.amazonaws.com`

File: pytorch/training/docker/2.0/py3/cu118/Dockerfile.gpu [hunk 2]
```diff
@@ -429,6 +429,17 @@ RUN /opt/conda/bin/mamba install -y python=${PYTHON_VERSION} pytorch=${PYTORCH_V
   -c nvidia \
   && /opt/conda/bin/mamba clean -afy
 
+# Override only sm-pytorch
+RUN /opt/conda/bin/mamba install --force-reinstall -y \
```
Comment (jeet4320) [inline pytorch/training/docker/2.0/py3/cu118/Dockerfile.gpu:433]: if you installing pytorch here, then this pytorch installation should be removed to not to increase DLC image size

https://github.com/aws/deep-learning-containers/pull/3566/files#diff-bc0b77b02ee5079e588e1fd8714585bee92e4159c585a980f2f576beffbc6228L421

Comment (access2rohit) [reply pytorch/training/docker/2.0/py3/cu118/Dockerfile.gpu:433]: I can add uninstall specifically for the torch package, we need the other packages installed

File: pytorch/training/docker/2.0/py3/cu121/Dockerfile.gpu [hunk 3]
```diff
@@ -352,26 +353,40 @@ ARG PYTORCH_VERSION_EC2
...
+# Install sm PyTorch
+RUN /opt/conda/bin/mamba install -y python=${PYTHON_SHORT_VERSION} \
+  pytorch="2.0.1=sm_py3.10_cuda12.1*" \
+  pytorch-cuda=12.1 cuda-nvcc=12.1.* torchvision torchaudio torchtext \
+  --override-channels \
+  -c ${SMD_MODEL_PARALLEL_CONDA_CHANNEL} \
+  -c https://aws-ml-conda-ec2.s3.us-west-2.amazonaws.com \
+  -c pytorch \
+  -c numba/label/dev \
+  -c pytorch-nightly \
```
Comment (jeet4320) [inline pytorch/training/docker/2.0/py3/cu121/Dockerfile.gpu:380]: why nightly is needed? does it have stable packages?

Comment (access2rohit) [reply pytorch/training/docker/2.0/py3/cu121/Dockerfile.gpu:380]: for torchtrition which is a dependency for sm pytorch

**Comment by jeet4320** [comment]
FYI: forch pushing or rebasing removes earlier CI on each commit. So merging is preferred

**Comment by jeet4320** [comment]
I see all EC2 and SM tests are getting skipped, looking. We should run all EC2 tests.

**Pull request closed without merging**